src/response.rs

```diff
@@ -595,24 +590,15 @@ impl Response {
             ));
         }
 
-        let compression =
-            get_header(&headers, "content-encoding").and_then(Compression::from_header_value);
-
-        let connection_option =
-            Self::connection_option(http_version, get_header(&headers, "connection"));
-
-        let body_type = Self::body_type(&unit.method, status, http_version, &headers);
-
-        // remove Content-Encoding and length due to automatic decompression
-        if compression.is_some() {
-            headers.retain(|h| !h.is_name("content-encoding") && !h.is_name("content-length"));
+        if consume {
+            stream.fill_buf()?;
+            stream.consume(tot_bytes_read);
         }
 
-        let reader =
-            Self::stream_to_reader(stream, &unit, body_type, compression, connection_option);
-
         let url = unit.url.clone();
 
+        let reader = Box::new(Cursor::new(vec![]));
```

Comment:
This extra allocation on every (regular, also not expect-100) call irks me a bit.

What if we replace `reader: Box<dyn Read + Send + Sync + 'static>` in `Response` with an enum `PendingReader` that has two values which explains the state.

`PendingReader::BeforeBodyStart` and `PendingReader::Reader(Box<dyn …>)`

Could be modeled as an `Option` too, but maybe a bit clearer with an enum. What do you think?

Reply:
I thought about Option too, but yeah it's more clear with enum so I went with that.